### DISCUSSION TITLE
Error with 1D Convolution layer.

### DISCUSSION THREAD

**Discussion by satishkumar538** [Xilinx/finn#418] [Q&A] [answered]
![Satish_Finn](https://user-images.githubusercontent.com/36513144/141070896-21b2c2d0-2814-4c92-9d1c-a8cb9159ec3a.png)
I am working a model having 1D Convolution layer with In_Channel=2, Out_Channel=16, Kernel=8. While build I am getting following Error (See Screenshot). Please suggest.

**Comment by fpjentzsch** [comment, marked as answer]
Hi,

to work with true QuantConv1d layers, you need to apply the following additional transformation prior to streamlining:
https://github.com/Xilinx/finn-base/blob/dev/src/finn/transformation/change_3d_tensors_to_4d.py
 
For an example how this can be done with a custom step, see our VGG10-RadioML example:
https://github.com/Xilinx/finn-examples/blob/main/build/vgg10-radioml/build.py
https://github.com/Xilinx/finn-examples/blob/main/build/vgg10-radioml/custom_steps.py

You could also add it directly in the default `step_streamline`:
https://github.com/Xilinx/finn/blob/dev/src/finn/builder/build_dataflow_steps.py#L245

**Background:**
Note that FINN still uses 4D (NHWC) tensors and 2D attributes (e.g. kernel_shape=(3,1)) internally, even for the 1D case. This means that you can either define your topology with QuantConv2d layers with H or W set to 1, or use proper QuantConv1d (etc.) layers and apply the Change3DTo4DTensors() transformation from finn-base before streamlining.
_(taken from this Gitter discussion: https://gitter.im/xilinx-finn/community?at=60a50036ee77a74d685ea9ae)_

We will consider adding this (relatively new) transformation to the default build flow in the future.
